nomos-da/network/core/src/swarm/executor.rs

```diff
@@ -0,0 +1,174 @@
+    },
+    replication::behaviour::ReplicationEvent,
+    sampling::behaviour::SamplingEvent,
+};
+use crate::swarm::common::{
+    handle_replication_event, handle_sampling_event, handle_validator_dispersal_event,
+};
+use crate::SubnetworkId;
+use subnetworks_assignations::MembershipHandler;
+
+pub struct ExecutorEventsStream {
+    pub sampling_events_receiver: UnboundedReceiverStream<SamplingEvent>,
+    pub validation_events_receiver: UnboundedReceiverStream<DaBlob>,
+    pub dispersal_events_receiver: UnboundedReceiverStream<DispersalExecutorEvent>,
+}
+
+pub struct ExecutorSwarm<
+    Membership: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId> + 'static,
+> {
+    swarm: Swarm<ExecutorBehaviour<Membership>>,
+    sampling_events_sender: UnboundedSender<SamplingEvent>,
+    validation_events_sender: UnboundedSender<DaBlob>,
+    dispersal_events_sender: UnboundedSender<DispersalExecutorEvent>,
+}
+
+impl<Membership> ExecutorSwarm<Membership>
+where
+    Membership: MembershipHandler<NetworkId = SubnetworkId, Id = PeerId> + Clone + Send,
+{
+    pub fn new(
+        key: Keypair,
+        membership: Membership,
+        addresses: AddressBook,
+    ) -> (Self, ExecutorEventsStream) {
+        let (sampling_events_sender, sampling_events_receiver) = unbounded_channel();
+        let (validation_events_sender, validation_events_receiver) = unbounded_channel();
+        let (dispersal_events_sender, dispersal_events_receiver) = unbounded_channel();
+        let sampling_events_receiver = UnboundedReceiverStream::new(sampling_events_receiver);
+        let validation_events_receiver = UnboundedReceiverStream::new(validation_events_receiver);
+        let dispersal_events_receiver = UnboundedReceiverStream::new(dispersal_events_receiver);
+        (
+            Self {
+                swarm: Self::build_swarm(key, membership, addresses),
+                sampling_events_sender,
+                validation_events_sender,
+                dispersal_events_sender,
+            },
+            ExecutorEventsStream {
+                sampling_events_receiver,
+                validation_events_receiver,
+                dispersal_events_receiver,
+            },
+        )
+    }
+    fn build_swarm(
+        key: Keypair,
+        membership: Membership,
+        addresses: AddressBook,
+    ) -> Swarm<ExecutorBehaviour<Membership>> {
+        SwarmBuilder::with_existing_identity(key)
+            .with_tokio()
+            .with_quic()
+            .with_behaviour(|key| ExecutorBehaviour::new(key, membership, addresses))
+            .expect("Validator behaviour should build")
+            .with_swarm_config(|cfg| {
+                cfg.with_idle_connection_timeout(Duration::from_secs(u64::MAX))
+            })
+            .build()
+    }
+
+    pub fn dial(&mut self, addr: Multiaddr) -> Result<(), DialError> {
+        self.swarm.dial(addr)?;
+        Ok(())
+    }
+
+    pub fn local_peer_id(&self) -> &PeerId {
```

Comment:
`dial` and `local_peer_id` are a wrapper functions for interacting with the libp2p swarm. I think it's a good decision because we might tweak this logic without changing the backend. In the same sense we could expose `blob_sender` and `open_stream_sender` channels, they'll be needed by the executor also.